# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: ipc/glue/NodeController.cpp
# Commit: 451508713a62
# Full Hash: 451508713a625beaaa9fcbbf06c65ca312dc34fb
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2021-06-23 09:53:24
# Regressor Bug: 1706374
# File Overlap Count: 2
# Description:
#   Bug 1706374 - Part 13: Remove the event footer from an IPC::Message when deserializing, r=handyman,glandium
#   
#   This unfortunately requires a new method to be added to BufferList to
#   support truncating the buffer to a particular iterator.
#   
# ==============================================================================

diff -r 878722e87e62 -r 451508713a62 ipc/glue/NodeController.cpp
--- a/ipc/glue/NodeController.cpp	Tue Jun 22 18:17:23 2021 +0000
+++ b/ipc/glue/NodeController.cpp	Tue Jun 22 18:17:23 2021 +0000
@@ -177,7 +177,8 @@
   MOZ_ASSERT(message->FooterSize() == aEvent->GetSerializedSize());
   Vector<char, 256, InfallibleAllocPolicy> buffer2;
   (void)buffer2.initLengthUninitialized(message->FooterSize());
-  MOZ_ASSERT(message->ReadFooter(buffer2.begin(), buffer2.length()));
+  MOZ_ASSERT(message->ReadFooter(buffer2.begin(), buffer2.length(),
+                                 /* truncate */ false));
   MOZ_ASSERT(!memcmp(buffer2.begin(), buffer.begin(), buffer.length()));
 #endif
 
@@ -188,7 +189,11 @@
     -> UniquePtr<Event> {
   Vector<char, 256, InfallibleAllocPolicy> buffer;
   (void)buffer.initLengthUninitialized(aMessage->FooterSize());
-  if (!aMessage->ReadFooter(buffer.begin(), buffer.length())) {
+  // Truncate the message when reading the footer, so that the extra footer data
+  // is no longer present in the message. This allows future code to eventually
+  // send the same `IPC::Message` to another process.
+  if (!aMessage->ReadFooter(buffer.begin(), buffer.length(),
+                            /* truncate */ true)) {
     NODECONTROLLER_WARNING("Call to ReadFooter for message '%s' Failed",
                            aMessage->name());
     return nullptr;